runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -240,6 +282,42 @@ void registerTable(@Nonnull String namespace,
         }
     }
 
+    /**
+     * For each protobuf filename -> descriptor map, validate if the previous descriptor matches
+     * If there is a new schema or an update, record the new update in the table.
+     *
+     * @param protoName - name of the protobuf file
+     * @param newProtoFd - descriptor of the protobuf file
+     * @param namespace - namespace of the table
+     * @param tableName - tablename without the namespace
+     * @return - true if schema change is detected, false otherwise
+     */
+    private boolean recordNewSchema(ProtobufFileName protoName,
+                                    CorfuRecord<ProtobufFileDescriptor, TableMetadata> newProtoFd,
+                                    String namespace,
+                                    String tableName) {
+        final CorfuRecord<ProtobufFileDescriptor, TableMetadata> oldProto = this.protobufDescriptorTable.get(protoName);
+        if (oldProto == null) {
+            this.protobufDescriptorTable.put(protoName, newProtoFd);
+            return false;
+        }
+        if (!oldProto.getPayload().getFileDescriptor().equals(newProtoFd.getPayload().getFileDescriptor())) {
+            if (!protoName.getFileName().startsWith("google/protobuf")) {
+                log.warn("registerTable: Schema update detected for table {}${} in schema file {}",
+                        namespace, tableName, protoName);
+
+                for (StackTraceElement st : Thread.currentThread().getStackTrace()) {
+                    log.debug("{}", st);
+                }
+                log.debug("registerTable: old schema: {}", oldProto.getPayload().getFileDescriptor());
+                log.debug("registerTable: new schema: {}", newProtoFd.getPayload().getFileDescriptor());
+            }
+            this.protobufDescriptorTable.put(protoName, newProtoFd);
```

Comment:
Should this fail instead of emitting a warning and just overriding the old schema?

Reply:
We do want to always override the schema that is how developers perform build-to-build upgrades.
Also overwriting schema is normal workflow during upgrades.

Reply:
Also please indicate in a comment that `recordNewSchema` should be execute in a transaction, otherwise an old schema and override a new one in certain race conditions. Also, how do you know what is old and what is new? Is there a monotonic counter that tracks this?

Reply:
done. existing means old, incoming means new.